packages/playwright-test/src/mount.ts

```diff
@@ -21,51 +21,47 @@ let boundCallbacksForMount: Function[] = [];
-  PlaywrightWorkerArgs & PlaywrightWorkerOptions & { _ctWorker: { page: Page | undefined, context: BrowserContext | undefined, hash: string, isolateTests: boolean } },
+  PlaywrightWorkerArgs & PlaywrightWorkerOptions & { _ctWorker: { context: BrowserContext | undefined, hash: string } },
   { _contextFactory: (options?: BrowserContextOptions) => Promise<BrowserContext> }> = {
 
-    _ctWorker: [{ page: undefined, context: undefined, hash: '', isolateTests: false }, { scope: 'worker' }],
-
-    context: async ({ _contextFactory, playwright, browser, _ctWorker, video, trace, viewport }, use, testInfo) => {
-      _ctWorker.isolateTests = shouldCaptureVideo(normalizeVideoMode(video), testInfo) || shouldCaptureTrace(normalizeTraceMode(trace), testInfo);
-      if (_ctWorker.isolateTests) {
-        await use(await _contextFactory());
-        return;
-      }
+    _ctWorker: [{ context: undefined, hash: '' }, { scope: 'worker' }],
 
+    context: async ({ playwright, browser, _ctWorker, _contextFactory, video, trace }, use, testInfo) => {
+      const isolateTests = shouldCaptureVideo(normalizeVideoMode(video), testInfo) || shouldCaptureTrace(normalizeTraceMode(trace), testInfo);
       const defaultContextOptions = (playwright.chromium as any)._defaultContextOptions as BrowserContextOptions;
       const hash = contextHash(defaultContextOptions);
 
-      if (!_ctWorker.page || _ctWorker.hash !== hash) {
+      if (!_ctWorker.context || _ctWorker.hash !== hash || isolateTests) {
         if (_ctWorker.context)
           await _ctWorker.context.close();
-
-        const context = await browser.newContext();
-        const page = await createPage(context);
-        _ctWorker.context = context;
-        _ctWorker.page = page;
+        // Context factory sets up video so we want to use that for isolated contexts.
+        // However, it closes the context after the test, so we don't want to use it
+        // for shared contexts.
+        _ctWorker.context = isolateTests ? await _contextFactory() : await browser.newContext();
         _ctWorker.hash = hash;
-        await use(page.context());
-        return;
+        await _ctWorker.context.addInitScript('navigator.serviceWorker.register = () => {}');
+        await _ctWorker.context.exposeFunction('__pw_dispatch', (ordinal: number, args: any[]) => {
+          boundCallbacksForMount[ordinal](...args);
+        });
       } else {
-        const page = _ctWorker.page;
-        await (page as any)._wrapApiCall(async () => {
-          await (page as any)._resetForReuse();
-          await (page.context() as any)._resetForReuse();
-          await page.goto('about:blank');
-          await page.setViewportSize(viewport || { width: 1280, height: 800 });
-          await page.goto(process.env.PLAYWRIGHT_VITE_COMPONENTS_BASE_URL!);
-        }, true);
-        await use(page.context());
+        await (_ctWorker.context as any)._resetForReuse();
```

Comment:
This will show up in tracing now.

It feels like starting with ToT and moving addInitScript/exposeFunction into context would already be as good as what you are doing here.

Reply:
Actually, this is now mutually exclusive w/ tracing, so all good!